// @alpacahq/alpaca-trade-api/dist/resources/account.js
'use strict';
var _a = require('lodash'), omitBy = _a.omitBy, isNil = _a.isNil;
function get() {
    return this.httpRequest('/account');
}
function updateConfigs(configs) {
    return this.httpRequest('/account/configurations', null, configs, 'PATCH');
}
function getConfigs() {
    return this.httpRequest('/account/configurations');
}
function getActivities(_a) {
    var activityTypes = _a.activityTypes, until = _a.until, after = _a.after, direction = _a.direction, date = _a.date, pageSize = _a.pageSize, pageToken = _a.pageToken;
    if (Array.isArray(activityTypes)) {
        activityTypes = activityTypes.join(",");
    }
    var queryParams = omitBy({
        activity_types: activityTypes,
        until: until,
        after: after,
        direction: direction,
        date: date,
        page_size: pageSize,
        page_token: pageToken
    }, isNil);
    return this.httpRequest('/account/activities', queryParams);
}
function getPortfolioHistory(_a) {
    var date_start = _a.date_start, date_end = _a.date_end, period = _a.period, timeframe = _a.timeframe, extended_hours = _a.extended_hours;
    var queryParams = omitBy({
        date_start: date_start,
        date_end: date_end,
        period: period,
        timeframe: timeframe,
        extended_hours: extended_hours
    }, isNil);
    return this.httpRequest('/account/portfolio/history', queryParams);
}
module.exports = {
    get: get,
    getConfigs: getConfigs,
    updateConfigs: updateConfigs,
    getActivities: getActivities,
    getPortfolioHistory: getPortfolioHistory
};
