// @alpacahq/alpaca-trade-api/dist/resources/datav2/stock_websocket_v2.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlpacaStocksClient = void 0;
var entityv2_1 = require("./entityv2");
var websocket_1 = require("./websocket");
var AlpacaStocksClient = /** @class */ (function (_super) {
    __extends(AlpacaStocksClient, _super);
    function AlpacaStocksClient(options) {
        var _this = this;
        var url = "wss" +
            options.url.substr(options.url.indexOf(":")) +
            "/v2/" +
            options.feed;
        options.url = url;
        options.subscriptions = {
            trades: [],
            quotes: [],
            bars: [],
            dailyBars: [],
            statuses: [],
            lulds: [],
        };
        _this = _super.call(this, options) || this;
        return _this;
    }
    AlpacaStocksClient.prototype.subscribeForTrades = function (trades) {
        var _a;
        (_a = this.session.subscriptions.trades).push.apply(_a, trades);
        this.subscribe({ trades: trades });
    };
    AlpacaStocksClient.prototype.subscribeForQuotes = function (quotes) {
        var _a;
        (_a = this.session.subscriptions.quotes).push.apply(_a, quotes);
        this.subscribe({ quotes: quotes });
    };
    AlpacaStocksClient.prototype.subscribeForBars = function (bars) {
        var _a;
        (_a = this.session.subscriptions.bars).push.apply(_a, bars);
        this.subscribe({ bars: bars });
    };
    AlpacaStocksClient.prototype.subscribeForDailyBars = function (dailyBars) {
        var _a;
        (_a = this.session.subscriptions.dailyBars).push.apply(_a, dailyBars);
        this.subscribe({ dailyBars: dailyBars });
    };
    AlpacaStocksClient.prototype.subscribeForStatuses = function (statuses) {
        var _a;
        (_a = this.session.subscriptions.statuses).push.apply(_a, statuses);
        this.subscribe({ statuses: statuses });
    };
    AlpacaStocksClient.prototype.subscribeForLulds = function (lulds) {
        var _a;
        (_a = this.session.subscriptions.lulds).push.apply(_a, lulds);
        this.subscribe({ lulds: lulds });
    };
    AlpacaStocksClient.prototype.subscribe = function (symbols) {
        var _a, _b, _c, _d, _e, _f;
        var subMsg = {
            action: "subscribe",
            trades: (_a = symbols.trades) !== null && _a !== void 0 ? _a : [],
            quotes: (_b = symbols.quotes) !== null && _b !== void 0 ? _b : [],
            bars: (_c = symbols.bars) !== null && _c !== void 0 ? _c : [],
            dailyBars: (_d = symbols.dailyBars) !== null && _d !== void 0 ? _d : [],
            statuses: (_e = symbols.statuses) !== null && _e !== void 0 ? _e : [],
            lulds: (_f = symbols.lulds) !== null && _f !== void 0 ? _f : [],
        };
        this.conn.send(this.msgpack.encode(subMsg));
    };
    AlpacaStocksClient.prototype.subscribeAll = function () {
        var _a = this.session.subscriptions, trades = _a.trades, quotes = _a.quotes, bars = _a.bars, dailyBars = _a.dailyBars, statuses = _a.statuses, lulds = _a.lulds;
        if (trades.length > 0 ||
            quotes.length > 0 ||
            bars.length > 0 ||
            dailyBars.length > 0 ||
            statuses.length > 0 ||
            lulds.level > 0) {
            var msg = {
                action: "subscribe",
                trades: trades,
                quotes: quotes,
                bars: bars,
                dailyBars: dailyBars,
                statuses: statuses,
                lulds: lulds,
            };
            this.conn.send(this.msgpack.encode(msg));
        }
    };
    AlpacaStocksClient.prototype.unsubscribeFromTrades = function (trades) {
        this.session.subscriptions.trades =
            this.session.subscriptions.trades.filter(function (trade) { return !trades.includes(trade); });
        this.unsubscribe({ trades: trades });
    };
    AlpacaStocksClient.prototype.unsubscribeFromQuotes = function (quotes) {
        this.session.subscriptions.quotes =
            this.session.subscriptions.quotes.filter(function (quote) { return !quotes.includes(quote); });
        this.unsubscribe({ quotes: quotes });
    };
    AlpacaStocksClient.prototype.unsubscribeFromBars = function (bars) {
        this.session.subscriptions.bars = this.session.subscriptions.bars.filter(function (bar) { return !bars.includes(bar); });
        this.unsubscribe({ bars: bars });
    };
    AlpacaStocksClient.prototype.unsubscriceFromDailyBars = function (dailyBars) {
        this.session.subscriptions.dailyBars =
            this.session.subscriptions.dailyBars.filter(function (dailyBar) { return !dailyBars.includes(dailyBar); });
        this.unsubscribe({ dailyBars: dailyBars });
    };
    AlpacaStocksClient.prototype.unsubscribeFromStatuses = function (statuses) {
        this.session.subscriptions.statuses =
            this.session.subscriptions.statuses.filter(function (status) { return !statuses.includes(status); });
        this.unsubscribe({ statuses: statuses });
    };
    AlpacaStocksClient.prototype.unsubscribeFromLulds = function (lulds) {
        this.session.subscriptions.statuses =
            this.session.subscriptions.lulds.filter(function (luld) { return !lulds.includes(luld); });
        this.unsubscribe({ lulds: lulds });
    };
    AlpacaStocksClient.prototype.unsubscribe = function (symbols) {
        var _a, _b, _c, _d, _e, _f;
        var unsubMsg = {
            action: "unsubscribe",
            trades: (_a = symbols.trades) !== null && _a !== void 0 ? _a : [],
            quotes: (_b = symbols.quotes) !== null && _b !== void 0 ? _b : [],
            bars: (_c = symbols.bars) !== null && _c !== void 0 ? _c : [],
            dailyBars: (_d = symbols.dailyBars) !== null && _d !== void 0 ? _d : [],
            statuses: (_e = symbols.statuses) !== null && _e !== void 0 ? _e : [],
            lulds: (_f = symbols.lulds) !== null && _f !== void 0 ? _f : [],
        };
        this.conn.send(this.msgpack.encode(unsubMsg));
    };
    AlpacaStocksClient.prototype.updateSubscriptions = function (msg) {
        this.log("listening to streams:\n        trades: " + msg.trades + ",\n        quotes: " + msg.quotes + ",\n        bars: " + msg.bars + ",\n        dailyBars: " + msg.dailyBars + ",\n        statuses: " + msg.statuses + ",\n        lulds: " + msg.lulds);
        this.session.subscriptions = {
            trades: msg.trades,
            quotes: msg.quotes,
            bars: msg.bars,
            dailyBars: msg.dailyBars,
            statuses: msg.statuses,
            lulds: msg.lulds,
        };
    };
    AlpacaStocksClient.prototype.onStockTrade = function (fn) {
        this.on(websocket_1.EVENT.TRADES, function (trade) { return fn(trade); });
    };
    AlpacaStocksClient.prototype.onStockQuote = function (fn) {
        this.on(websocket_1.EVENT.QUOTES, function (quote) { return fn(quote); });
    };
    AlpacaStocksClient.prototype.onStockBar = function (fn) {
        this.on(websocket_1.EVENT.BARS, function (bar) { return fn(bar); });
    };
    AlpacaStocksClient.prototype.onStockDailyBar = function (fn) {
        this.on(websocket_1.EVENT.DAILY_BARS, function (dailyBar) { return fn(dailyBar); });
    };
    AlpacaStocksClient.prototype.onStatuses = function (fn) {
        this.on(websocket_1.EVENT.TRADING_STATUSES, function (status) { return fn(status); });
    };
    AlpacaStocksClient.prototype.onLulds = function (fn) {
        this.on(websocket_1.EVENT.LULDS, function (luld) { return fn(luld); });
    };
    AlpacaStocksClient.prototype.dataHandler = function (data) {
        var _this = this;
        data.forEach(function (element) {
            if ("T" in element) {
                switch (element.T) {
                    case "t":
                        _this.emit(websocket_1.EVENT.TRADES, (0, entityv2_1.AlpacaTradeV2)(element));
                        break;
                    case "q":
                        _this.emit(websocket_1.EVENT.QUOTES, (0, entityv2_1.AlpacaQuoteV2)(element));
                        break;
                    case "b":
                        _this.emit(websocket_1.EVENT.BARS, (0, entityv2_1.AlpacaBarV2)(element));
                        break;
                    case "d":
                        _this.emit(websocket_1.EVENT.DAILY_BARS, (0, entityv2_1.AlpacaBarV2)(element));
                        break;
                    case "s":
                        _this.emit(websocket_1.EVENT.TRADING_STATUSES, (0, entityv2_1.AlpacaStatusV2)(element));
                        break;
                    case "l":
                        _this.emit(websocket_1.EVENT.LULDS, (0, entityv2_1.AlpacaLuldV2)(element));
                        break;
                    default:
                        _this.emit(websocket_1.EVENT.CLIENT_ERROR, websocket_1.ERROR.UNEXPECTED_MESSAGE);
                }
            }
        });
    };
    return AlpacaStocksClient;
}(websocket_1.AlpacaWebsocket));
exports.AlpacaStocksClient = AlpacaStocksClient;
